### PR TITLE
Structured load api

### PR DESCRIPTION
**Pull request by pp-mo**
This now replaces the discussion PR on the "feature branch" in my repo : https://github.com/pp-mo/iris/pull/27

**NOTE NOT READY YET :** documentation changes still need to be reviewed/finalised/agreed.

from notes there:
> As everything else was already agreed + merged to here, I've merged the outstanding [pp-mo/iris#31 "Revised documentation, whatsnew and deprecations."](https://github.com/pp-mo/iris/pull/#31)  here, as that is all that remains to discuss.
So I closed this and opened a new PR to the main Iris master instead : #2241
We can discuss what remains to be discussed there.

### CODE REVIEW

File: docs/iris/src/whatsnew/contributions_1.11/newfeature_2016-Nov-09_structured_um_loading.txt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  Fast UM file loading:
+  ---------------------
+  Support has been added for accelerated loading of UM files (PP and
+  Fieldsfile), when these have a suitable regular 'structured' form.
+  Normal Fieldsfiles are generally suitable, and many PP files also.
+
+  A context manager is used to enable fast um loading in all the regular iris
+  load functions, such as :meth:`iris.load` and :meth:`iris.load_cube`,
+  when loading data from UM file types.
+
```
Comment (pp-mo) [inline docs/iris/src/whatsnew/contributions_1.11/newfeature_2016-Nov-09_structured_um_loading.txt]: Add a trivial example of usage.

File: lib/iris/fileformats/um/_fast_load.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+    Use of load callbacks:
+
+        When a user callback function is used with structured-loading, it is
+        called in a somewhat different way than in a 'normal' load :
+        The callback is called once for each basic *structured* cube loaded,
+        which is normally the whole of one phenomenon from a single input file.
+        In particular, the callback's "field" argument is a
+        :class:`~iris.fileformats.um.FieldCollation`, from which "field.fields"
+        gives a *list* of PPFields from which that cube was built.
+        The code required is therefore different from a 'normal' callback.
```
Comment (pp-mo) [inline lib/iris/fileformats/um/_fast_load.py]: Add a link to the test code, providing example of "translating" a callback.

File: docs/iris/src/whatsnew/contributions_1.11/newfeature_2016-Nov-09_structured_um_loading.txt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  [[
+  **NOTE TO RELEASE COMPILER**
+  This needs highlighting, and a better explanation than the plain reference
+  documentation.  We probably want a "featured item" section in the whatsnew,
+  along the following lines ...
+  ]]
+  Fast UM file loading:
+  ---------------------
+  Support has been added for accelerated loading of UM files (PP and
```
Comment (pp-mo) [inline docs/iris/src/whatsnew/contributions_1.11/newfeature_2016-Nov-09_structured_um_loading.txt]: Add specific test-data example with speedup factor.

**Comment by pp-mo** [comment]
I think I've now addressed all the outstanding stuff we noted down.
Please re-review @marqh .

**Comment by marqh** [comment]
thank you for all the hard work @pp-mo 

I agree that this is good to go, we have captured a few 'nice to have' elements in #2243 
this code belongs on master now
any further changes will be additive

**Pull request merged**